### PR TITLE
Improved `read_redcap()` warnings for databases with access restrictions

### LINKED ISSUE #38: [FEATURE] Add test cases for attempting to access PHI-restricted data
# Feature Request Description
REDCap has settings to restrict export of fields identified as PHI. These settings impact the content of fields returned by `REDCapR`.
deidentified-redcap

Our testing suite should include tests that check for expected behavior when a user attempts to access a form for which they don't have full permissions to export.

# Proposed Solution
This is an initial pass at expected behavior across a range cases. The parameters I'm varying are:
1. Whether the form was requested specifically vs. requested by default
2. Whether the user has access to all, some, or no fields in the form

| Request | Permissions within form | Proposed Behavior | `REDCapR` Behavior |
|---|---|---|---|
| `forms = NULL` (default) | All fields |  success | success, all accessible fields returned |
| `forms = NULL` (default) | Subset of fields |  warning? | success, all accessible fields returned |
| `forms = NULL` (default) | No fields | warning? | success, all accessible fields returned |
| `forms = "name_of_form"` | All fields | success | success, all accessible fields returned |
| `forms = "name_of_form"` | Subset of fields | warning? | success, all accessible fields returned |
| `forms = "name_of_form"` | No fields | warning | success, all fields in the project returned *including ones the user doesn't have access to* |

### Checklist

- [x] The issue is atomic
- [x] The issue description is documented
- [x] The issue title describes the problem succinctly
- [x] Developers are assigned to the issue
- [x] Labels are assigned to the issue

### PR DESCRIPTION
**Pull request by ezraporter**
# Description
This PR implements an improved scheme for issuing warnings/errors when a user has restricted access to data in the REDCap project. There are 3 cases to consider:

1. User gets back no data from any instruments because of PHI/de-identified or user doesn't have access to any requested instrument
    - User gets an error saying no data was returned
2. User gets back partial data for an instrument because of PHI/de-identified data
    - User gets a warning listing the missing fields
3. User gets back no data for an instrument because all data is PHI/de-identified or user doesn't have access to the instrument
    - User gets a warning that no data was returned and the instrument won't appear in the output

# Proposed Changes 
List changes below in bullet format:

- Update `check_user_rights()` to error when none of the fields in the metadata are present in the data
- Update `check_user_rights()` to issue more specific warning messages when an instrument returns no or partial data
- Add new `REDCAPTIDIER_RESTRICTED_ACCESS_API` test REDCap + mocks and add test cases for restricted access

**Screenshots**
error

warning

### Issue Addressed
closes #38

### PR Checklist
Before submitting this PR, please check and verify below that the submission meets the below criteria:

- [x] New/revised functions have associated tests
- [x] New/revised functions that update downstream outputs have associated static testing files (`.RDS`) updated under `inst/testdata/create_test_data.R`
- [x] New tests that make API calls use `httptest::with_mock_api` and any new mocks were added to `tests/testthat/fixtures/create_httptest_mocks.R`
- [x] New/revised functions use appropriate naming conventions
- [x] New/revised functions don't repeat code
- [x] Code changes are less than **250** lines total
- [x] Issues linked to the PR using [GitHub's list of keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] The appropriate reviewer is assigned to the PR
- [x] The appropriate developers are assigned to the PR
- [x] Pre-release package version incremented using `usethis::use_dev_version()`

# Code Review
This section to be used by the reviewer and developers during Code Review after PR submission

### Code Review Checklist

- [ ] I checked that new files follow naming conventions and are in the right place
- [ ] I checked that documentation is complete, clear, and without typos
- [ ] I added/edited comments to explain "why" not "how"
- [ ] I checked that all new variable and function names follow naming conventions
- [ ] I checked that new tests have been written for key business logic and/or bugs that this PR fixes
- [ ] I checked that new tests address important edge cases

### CODE REVIEW

**Comment by ezraporter** [comment]
CI seems to be failing due to issues connecting to the Azure Ubuntu Archive:

`Could not connect to azure.archive.ubuntu.com:80 (52.252.75.106), connection timed out`

They seem to pass when rerunning but take longer than usual. May be an issue (hopefully temporary) with Azure or GitHub. Folks seem to be discussing here: https://github.com/actions/runner-images/issues/675

File: R/checks.R [hunk 1]
```diff
@@ -24,20 +25,63 @@
...
+
+  db_fields <- names(db_data)
+  # Fields in metadata excluding record ID
+  metadata_fields <- setdiff(db_metadata$field_name_updated, get_record_id_field(db_data))
+
+  # If no data fields were returned the user has no data access
+  if (length(intersect(db_fields, metadata_fields)) == 0) {
+    cli_abort(
+      c(
+        "x" = "The REDCap API returned no data for any of the requested instruments.",
```
Comment (rsh52) [inline R/checks.R:41]: I think we should replace all messages that say "instruments" with "forms" since the argument is `forms =` and it's also the column name in the output super tibble (`redcap_form_name`).

File: R/checks.R [hunk 2]
```diff
@@ -24,20 +25,63 @@
...
+        cli_text("Instrument {.code {form_name}} is missing {qty(missing_fields)} field{?s}
+                 {.code {missing_fields}}.") |>
+          cli_fmt(collapse = TRUE, strip_newline = TRUE)
+      }
+    })
+
+  names(msg) <- rep("!", length(msg))
+
+  msg <- c(
+    "!" = "{qty(missing_db_metadata$form_name)}{?An/Some} instrument{?s} exist{?s/} in the
```
Comment (rsh52) [inline R/checks.R:76]: See instrument/form comment above

**Review by rsh52 — CHANGES_REQUESTED** [review]
Logic is all sound, a few comments about language and syntax.

File: R/checks.R [hunk 3]
```diff
@@ -24,20 +25,63 @@
...
+      class = c("no_data_access", cond_class),
+      call = call
+    )
+  }
+
+  missing_db_metadata <- db_metadata %>%
     select("field_name_updated", "form_name") %>%
     group_by(.data$form_name) %>%
-    summarise(fields = list(.data$field_name_updated))
+    filter(!all(.data$field_name_updated %in% db_fields)) |>
```
[the remark this answers is not in the document]
Comment (ezraporter) [reply R/checks.R:52]: Ooo good catch. Fixing this.

I did a find all to make sure and it looks like the only place we have `|>` is in the vignettes. Does that matter? Should we switch to `%>%` there to be safe?

File: R/checks.R (hunk 1 above)
Comment (ezraporter) [reply R/checks.R:41]: I'm fine with that. My only hesitation is that we seem to prefer "instrument" in the glossary: https://chop-cgtinformatics.github.io/REDCapTidieR/articles/glossary.html#instrument

File: R/checks.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (rsh52) [reply R/checks.R:52]: It probably doesn't matter for the vignettes, since those will generally show REDCapTidieR in use with other things, but I remember we talked about keeping `%>%` for a while yet so that we don't have to force users to rely on newer versions of R.

File: R/checks.R (hunk 1 above)
Comment (rsh52) [reply R/checks.R:41]: Yea... I also saw that. My opinion is still that if `forms` is used as the argument, and the error points users to their argument, then it should reflect that. Thoughts from @skadauke?

**Comment by skadauke** [comment]
I think using |> in the vignettes is good style. Using %>% in the package is good for backwards compatibility.
________________________________
From: ezraporter ***@***.***>
Sent: Monday, February 6, 2023 4:19 PM
To: CHOP-CGTInformatics/REDCapTidieR ***@***.***>
Cc: Stephan Kadauke ***@***.***>; Review requested ***@***.***>
Subject: [External]Re: [CHOP-CGTInformatics/REDCapTidieR] Improved `read_redcap()` warnings for databases with access restrictions (PR #123)


@ezraporter commented on this pull request.

________________________________

In R/checks.R<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTInformatics%2FREDCapTidieR%2Fpull%2F123%23discussion_r1097930375&data=05%7C01%7Ckadaukes%40chop.edu%7C602a2f13c40a485c84e908db0887cfe9%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638113151623920248%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=4QqGGjSqdT3nrfSeftMU28LQhbI5xv5%2F3Eg0BggzRUw%3D&reserved=0>:

>      select("field_name_updated", "form_name") %>%
     group_by(.data$form_name) %>%
-    summarise(fields = list(.data$field_name_updated))
+    filter(!all(.data$field_name_updated %in% db_fields)) |>


Ooo good catch. Fixing this.

I did a find all to make sure and it looks like the only place we have |> is in the vignettes. Does that matter? Should we switch to %>% there to be safe?

—
Reply to this email directly, view it on GitHub<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTInformatics%2FREDCapTidieR%2Fpull%2F123%23discussion_r1097930375&data=05%7C01%7Ckadaukes%40chop.edu%7C602a2f13c40a485c84e908db0887cfe9%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638113151623920248%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=4QqGGjSqdT3nrfSeftMU28LQhbI5xv5%2F3Eg0BggzRUw%3D&reserved=0>, or unsubscribe<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACTGHWQWELV4ZCQQ5BBRUDLWWFTFLANCNFSM6AAAAAAUS4YEEA&data=05%7C01%7Ckadaukes%40chop.edu%7C602a2f13c40a485c84e908db0887cfe9%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638113151623920248%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=X%2BoiHk5G26VS2%2B84yBbZNvfxB34F5kML2J%2BG83qjZv0%3D&reserved=0>.
You are receiving this because your review was requested.Message ID: ***@***.***>

** This email originated from an EXTERNAL sender to CHOP. Proceed with caution when replying, opening attachments, or clicking links. Do not disclose your CHOP credentials, employee information, or protected health information to a potential hacker**.

**Comment by skadauke** [comment]
I think we want to stick to "instruments" in the messages since that word has a more specific meaning in the context of REDCap than "forms". We use "forms" in the code since the REDCap API uses "forms".

From the Glossary:

[cid:d28b5724-c0df-4b9c-92be-7396a6e4a55b]
________________________________
From: Rich Hanna ***@***.***>
Sent: Monday, February 6, 2023 4:07 PM
To: CHOP-CGTInformatics/REDCapTidieR ***@***.***>
Cc: Stephan Kadauke ***@***.***>; Review requested ***@***.***>
Subject: [External]Re: [CHOP-CGTInformatics/REDCapTidieR] Improved `read_redcap()` warnings for databases with access restrictions (PR #123)


@rsh52 requested changes on this pull request.

Logic is all sound, a few comments about language and syntax.

________________________________

In R/checks.R<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTInformatics%2FREDCapTidieR%2Fpull%2F123%23discussion_r1097911165&data=05%7C01%7Ckadaukes%40chop.edu%7C250eb80dcfca4770662708db08862476%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638113144480203709%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=RbKCygHxY3jgF9IHpcgQ%2Blb6zaIJwW2AxHdX5duH9VA%3D&reserved=0>:

>      select("field_name_updated", "form_name") %>%
     group_by(.data$form_name) %>%
-    summarise(fields = list(.data$field_name_updated))
+    filter(!all(.data$field_name_updated %in% db_fields)) |>


Unsure if I've missed this in other PRs, but I don't think we've updated this package to use base R pipes yet did we? We may want to double check since I see in the difference here that there's a mix of %>% and |>

________________________________

In R/checks.R<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTInformatics%2FREDCapTidieR%2Fpull%2F123%23discussion_r1097916604&data=05%7C01%7Ckadaukes%40chop.edu%7C250eb80dcfca4770662708db08862476%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638113144480203709%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=C8alT0K9udVVFgnxWf9yZHk3In5uw9aWOGDSK%2Bwr2QM%3D&reserved=0>:

> -  missing_db_metadata <- db_metadata %>% # nolint: object_usage_linter
-    filter(!.data$field_name_updated %in% names(db_data)) %>%
+  # Set condition components used in both error and warning
+  msg_info <- c("i" = "This can happen when the user privileges are not set to allow exporting certain
+    instruments via the API.")
+  cond_class <- c("redcap_user_rights", "REDCapTidieR_cond")
+
+  db_fields <- names(db_data)
+  # Fields in metadata excluding record ID
+  metadata_fields <- setdiff(db_metadata$field_name_updated, get_record_id_field(db_data))
+
+  # If no data fields were returned the user has no data access
+  if (length(intersect(db_fields, metadata_fields)) == 0) {
+    cli_abort(
+      c(
+        "x" = "The REDCap API returned no data for any of the requested instruments.",


I think we should replace all messages that say "instruments" with "forms" since the argument is forms = and it's also the column name in the output super tibble (redcap_form_name).

________________________________

In R/checks.R<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTInformatics%2FREDCapTidieR%2Fpull%2F123%23discussion_r1097916785&data=05%7C01%7Ckadaukes%40chop.edu%7C250eb80dcfca4770662708db08862476%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638113144480203709%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=AivJonY8URE8AtsDjHV3KRfXTBN9Jzko2AK8EX4qEeM%3D&reserved=0>:

> +    missing_db_metadata,
+    .f = function(form_name, missing_fields, all_fields_missing) {
+      if (all_fields_missing) {
+        cli_text("Instrument {.code {form_name}} returned no data and will be removed from the output.") |>
+          cli_fmt(collapse = TRUE, strip_newline = TRUE)
+      } else {
+        cli_text("Instrument {.code {form_name}} is missing {qty(missing_fields)} field{?s}
+                 {.code {missing_fields}}.") |>
+          cli_fmt(collapse = TRUE, strip_newline = TRUE)
+      }
+    })
+
+  names(msg) <- rep("!", length(msg))
+
+  msg <- c(
+    "!" = "{qty(missing_db_metadata$form_name)}{?An/Some} instrument{?s} exist{?s/} in the


See instrument/form comment above

—
Reply to this email directly, view it on GitHub<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FCHOP-CGTInformatics%2FREDCapTidieR%2Fpull%2F123%23pullrequestreview-1285999987&data=05%7C01%7Ckadaukes%40chop.edu%7C250eb80dcfca4770662708db08862476%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638113144480203709%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=rsoeGeaiAzWTLY07FJh2jdzTtbQLGv%2BLX6cGe3Z0gvA%3D&reserved=0>, or unsubscribe<https://nam10.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACTGHWWPYAJ3Y5I7BW4OO3TWWFRYPANCNFSM6AAAAAAUS4YEEA&data=05%7C01%7Ckadaukes%40chop.edu%7C250eb80dcfca4770662708db08862476%7Ca611241607b041a59bb1d146b575c975%7C0%7C0%7C638113144480203709%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=dQ%2Bakqbw8cCXP2ZhiKM7mJaeE360qhfA%2Bk2W0O5%2FbQY%3D&reserved=0>.
You are receiving this because your review was requested.Message ID: ***@***.***>

** This email originated from an EXTERNAL sender to CHOP. Proceed with caution when replying, opening attachments, or clicking links. Do not disclose your CHOP credentials, employee information, or protected health information to a potential hacker**.

**Review by rsh52 — APPROVED** [review]
Looks good! 🚀
